README.md
# Multipart Uploads using Laravel, AWS S3, and Uppy

[![Latest Version on Packagist](https://img.shields.io/packagist/v/tapp/laravel-uppy-s3-multipart-upload.svg?style=flat-square)](https://packagist.org/packages/tapp/laravel-uppy-s3-multipart-upload)
[![Total Downloads](https://img.shields.io/packagist/dt/tapp/laravel-uppy-s3-multipart-upload.svg?style=flat-square)](https://packagist.org/packages/tapp/laravel-uppy-s3-multipart-upload)

Upload large files directly to [AWS S3](https://aws.amazon.com/s3/) using [Laravel](https://laravel.com/) (backend) and [Uppy](https://uppy.io/) (frontend).

## Appearance

![upload00](https://raw.githubusercontent.com/TappNetwork/laravel-uppy-s3-multipart-upload/master/docs/upload00.png)

![upload01](https://raw.githubusercontent.com/TappNetwork/laravel-uppy-s3-multipart-upload/master/docs/upload01.png)

![upload02](https://raw.githubusercontent.com/TappNetwork/laravel-uppy-s3-multipart-upload/master/docs/upload02.png)

## Installation

### Install the package via Composer

```bash
composer require tapp/laravel-uppy-s3-multipart-upload
```

**Heads up!**

If you are using a fresh Laravel installation with its latest version, it uses PSR-7 version 2, and currently, the AWS SDK for PHP is incompatible with this version.

Please follow this issue for updates: https://github.com/aws/aws-sdk-php/issues/2264

In the meantime, to install this package in a fresh Laravel app, use the `--update-with-dependencies` option:

```bash
$ composer require tapp/laravel-uppy-s3-multipart-upload --update-with-dependencies
```

Also, to install the Flysystem S3 adapter:

```bash
$ composer require league/flysystem-aws-s3-v3:"^1.0"
```

### Add required JS libraries

Add on your `package.json` file the Uppy JS libraries and AlpineJS library:

```
    ...
    "devDependencies": {
        "alpinejs": "^2.7.3",
        ...
    },
    "dependencies": {
        "@uppy/aws-s3-multipart": "^2.0.2",
        "@uppy/core": "^2.0.2",
        "@uppy/drag-drop": "^2.0.1",
        "@uppy/status-bar": "^2.0.1"
        ...
    }
    ...
```

Add in your `resources/js/bootstrap.js` file:

```javascript
...

require('@uppy/core/dist/style.min.css')
require('@uppy/drag-drop/dist/style.min.css')
require('@uppy/status-bar/dist/style.min.css')

import Uppy from '@uppy/core'
import DragDrop from '@uppy/drag-drop'
import StatusBar from '@uppy/status-bar'
import AwsS3Multipart from '@uppy/aws-s3-multipart'

window.Uppy = Uppy
window.DragDrop = DragDrop
window.StatusBar = StatusBar
window.AwsS3Multipart = AwsS3Multipart
```

Add in your `resources/js/app.js`:

```javascript
...
require('alpinejs');
```

Install the JS libraries:

```
$ npm install
$ npm run dev
```

> You can use CDNs for [Uppy](https://uppy.io/docs/#With-a-script-tag) and [AlpineJS](https://github.com/alpinejs/alpine), if you prefer.

### Publish config file

Publish the config file with:
```bash
php artisan vendor:publish --tag=uppy-s3-multipart-upload-config
```

This is the contents of the published config file:

```php
return [
    's3' => [
        'bucket' => [
            /*
             * Folder on bucket to save the file
             */
            'folder' => '',
        ],
        'presigned_url' => [
            /*
             * Expiration time of the presigned URLs
             */
            'expiry_time' => '+1 hour',
        ],
    ],
];
```

### Publish view file

```bash
php artisan vendor:publish --tag=uppy-s3-multipart-upload-views
```

### AWS S3 Setup

This package installs the [AWS SDK for PHP](https://github.com/aws/aws-sdk-php) and use Laravel's default `s3` disk configuration from `config/filesystems.php` file.

You just have to add your S3 keys, region, and bucket using the following env vars in your `.env` file:

```
AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=
AWS_BUCKET=
AWS_URL="https://s3.amazonaws.com"
AWS_POST_END_POINT="https://${AWS_BUCKET}.s3.amazonaws.com/"
```

To allow direct multipart uploads to your S3 bucket, you need to add some extra configuration on bucket's `CORS configuration`.
On your AWS S3 console, select your bucket.
Click on `"Permissions"` tab.
On `"CORS configuration"` add the following configuration:

```
[
    {
        "AllowedHeaders": [
            "Authorization",
            "x-amz-date",
            "x-amz-content-sha256",
            "content-type"
        ],
        "AllowedMethods": [
            "PUT",
            "POST",
            "DELETE",
            "GET"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": [
            "ETag"
        ]
    }
]
```

On `AllowedOrigins`:

```
"AllowedOrigins": [
    "*"
]
```

You should list the URLs allowed, e.g.:

```
"AllowedOrigins": [
    "https://example.com"
]
```

https://uppy.io/docs/aws-s3-multipart/#S3-Bucket-Configuration

https://uppy.io/docs/aws-s3/#S3-Bucket-configuration

#### Add S3 Transfer Acceleration

To use [S3 transfer acceleration](https://docs.aws.amazon.com/AmazonS3/latest/userguide/transfer-acceleration.html),
enable it by adding a `AWS_USE_ACCELERATE_ENDPOINT=true` env var on your `.env` file, 
and add `'use_accelerate_endpoint' => env('AWS_USE_ACCELERATE_ENDPOINT')` in `s3` options on your `config/filesystems.php`:

```php
       's3' => [
            ...
            'use_accelerate_endpoint' => env('AWS_USE_ACCELERATE_ENDPOINT'),
        ],
```

#### Configuration

You can configure the folder to upload the files and the expiration of the presigned URLs used to upload the parts, with the `config/uppy-s3-multipart-upload.php` file:

```php
return [
    's3' => [
        'bucket' => [
            /*
             * Folder on bucket to save the file
             */
            'folder' => 'videos',
        ],
        'presigned_url' => [
            /*
             * Expiration time of the presigned URLs
             */
            'expiry_time' => '+30 minutes',
        ],
    ],
];
```

## Endpoints added

This package add the following routes:

```
POST    /s3/multipart
OPTIONS /s3/multipart
GET     /s3/multipart/{uploadId}
GET     /s3/multipart/{uploadId}/batch
POST    /s3/multipart/{uploadId}/complete
DELETE  /s3/multipart/{uploadId}
```

## Usage

### Add a hidden field for the uploaded file url

Add a hidden input form element on your blade template. When the upload is finished, it will receive the url of the uploaded file:

E.g.:

```html
<input type="hidden" name="file" id="file" />
```

### Add the `uppy` blade component to your blade view:

```html
<x-input.uppy />
```

### Passing data to the uppy blade component

**Hidden field name**

Use the `hiddenField` attribute to provide the name of the hidden field that will receive the url of uploaded file:

```php
$hiddenField = 'image_url';
```

```html
<x-input.uppy :hiddenField="$hiddenField" />
```

The `file` name will be used if none is provided.


**Uppy Core Options**

https://uppy.io/docs/uppy/#Options

You can pass any uppy options via `options` attribute:

```html
<x-input.uppy :options="$uppyOptions" />
```

Uppy core options are in this format:

```
$uppyOptions = "{
    debug: true,
    autoProceed: true,
    allowMultipleUploads: false,
}";
```

Default core options if none is provided:

```
{
    debug: true,
    autoProceed: true,
    allowMultipleUploads: false,
}
```

**Uppy Status Bar Options**

https://uppy.io/docs/status-bar/#Options

You can pass any uppy status bar options via `statusBarOptions` attribute:

```html
<x-input.uppy :statusBarOptions="$uppyStatusBarOptions" />
```

Uppy Status Bar options are in this format:

```
$uppyStatusBarOptions = "{
    target: '.upload .for-ProgressBar',
    hideAfterFinish: false,
}";
```

Default status bar options if none is provided:

```
{
    target: '.upload .for-ProgressBar',
    hideAfterFinish: false,
}
```

**Uppy Drag & Drop Options**

https://uppy.io/docs/drag-drop/#Options

You can pass any uppy drag & drop options via `dragDropOptions` attribute:

```html
<x-input.uppy :dragDropOptions="$uppyDragDropOptions" />
```

Uppy Drag & Drop options are in this format:

```
$uppyDragDropOptions = "{
    target: '.upload .for-DragDrop',
}";
```

Default drag & drop options if none is informed:

```
{
    target: '.upload .for-DragDrop',
}
```

**Upload Element Class**

Use the `uploadElementClass` attribute to provide the class of the HTML element used for upload:

```php
$imageClass = 'images';
```

```html
<x-input.uppy :uploadElementClass="$imageClass" />
```

The `upload` class will be used if none is provided.

**Multiple Uppy Instances**

If you want to use multiple Uppy instances, add a different `uploadElementClass` attribute to each instance. E.g.:

```html
<!-- First Uppy instance for image uploads -->
<div>
    <input type="hidden" name="images" id="images" />
    <x-input.uppy :options="$imageOptions" :hiddenField="$imageField" :uploadElementClass="$imageClass" />
</div>


<!-- Second Uppy instance for video uploads -->
<div>
    <input type="hidden" name="videos" id="videos" />
    <x-input.uppy :options="$videoOptions" :hiddenField="$videoField" :uploadElementClass="$videoClass" />
</div>
```

**Note from Uppy docs**: _"If multiple Uppy instances are being used, for instance, on two different pages,
an id should be specified. This allows Uppy to store information in localStorage without colliding with other Uppy instances."_
[Learn more here](https://uppy.io/docs/uppy/#id-39-uppy-39).

**Extra JavaScript to onUploadSuccess**

If you need to add extra JavaScript code on `onUploadSuccess` function, use the `extraJSForOnUploadSuccess` attribute:

E.g.:

```php
$extraJSForOnUploadSuccess = "
    document.getElementById('saveImageButton').removeAttribute('disabled');
    document.getElementById('saveImageButton').click();
"
```

```html
<x-input.uppy :extraJSForOnUploadSuccess="$extraJSForOnUploadSuccess" />
```

Default `extraJSForOnUploadSuccess` value is empty string.

### Clear caches

Run:

```
php artisan optimize
php artisan view:clear
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

If you discover any security-related issues, please email security@tappnetwork.com.

## Credits

- [Tapp Network](https://github.com/TappNetwork)
- [All Contributors](../../contributors)

### Libraries used in this package:

- [AWS SDK for PHP](https://github.com/aws/aws-sdk-php)
- [Uppy](https://uppy.io)
- [AlpineJS](https://github.com/alpinejs/alpine)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-uppy-s3-multipart-upload` will be documented in this file.

## 0.1 - 2021-02-11

- Initial release

## 0.2 - 2021-02-25

- Add configuration file
- Fix variable name
- Allow CSRF
- Update README

## 0.3 - 2021-02-26

- Add S3 transfer acceleration configuration
- Update README

## 0.3.1 - 2021-02-27

- Move S3 transfer acceleration config to the package config file

## 0.3.2 - 2021-02-28

- Update the way credentials are passed to use the Credentials class

## 0.3.3 - 2021-03-03

- Remove passing credentials, assuming that AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are set in the .env based on the AWS SDK suggestions

## 0.3.4 - 2021-03-18

- Use Flysystem AWS client

## 0.3.5 - 2021-09-15

- Add support for multiple Uppy instances
- Update README

## 0.4 - 2021-09-20

- Upgrade Uppy from 1.x to 2.0
- Add support for pre-sign URLs in batches
- Update README
- Add upgrade guide


src/Http/Controllers/UppyS3MultipartController.php
<?php

namespace Tapp\LaravelUppyS3MultipartUpload\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;
use Throwable;

class UppyS3MultipartController extends Controller
{
    protected $client;

    protected $bucket;

    public function __construct()
    {
        $this->client = Storage::disk('s3')->getDriver()->getAdapter()->getClient();

        $this->bucket = config('filesystems.disks.s3.bucket');
    }

    protected function encodeURIComponent($str)
    {
        if (!function_exists('encodeURIComponent')) {
            $revert = ['%21'=>'!', '%2A'=>'*', '%27'=>"'", '%28'=>'(', '%29'=>')', '%2F'=>'/'];

            return strtr(rawurlencode($str), $revert);
        }

        return encodeURIComponent($str);
    }

    public function createMultipartUploadOptions(Request $request)
    {
        header('Access-Control-Allow-Headers: Authorization, Content-Type, X-CSRF-TOKEN');

        return response([
            'message' => 'No content',
        ], 204);
    }

    /*
        S3 Syntax:

            https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#createmultipartupload
            https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html

            $result = $client->createMultipartUpload([
                'ACL' => 'private|public-read|public-read-write|authenticated-read|aws-exec-read|bucket-owner-read|bucket-owner-full-control',
                'Bucket' => '<string>', // REQUIRED
                'BucketKeyEnabled' => true || false,
                'CacheControl' => '<string>',
                'ContentDisposition' => '<string>',
                'ContentEncoding' => '<string>',
                'ContentLanguage' => '<string>',
                'ContentType' => '<string>',
                'ExpectedBucketOwner' => '<string>',
                'Expires' => <integer || string || DateTime>,
                'GrantFullControl' => '<string>',
                'GrantRead' => '<string>',
                'GrantReadACP' => '<string>',
                'GrantWriteACP' => '<string>',
                'Key' => '<string>', // REQUIRED
                'Metadata' => ['<string>', ...],
                'ObjectLockLegalHoldStatus' => 'ON|OFF',
                'ObjectLockMode' => 'GOVERNANCE|COMPLIANCE',
                'ObjectLockRetainUntilDate' => <integer || string || DateTime>,
                'RequestPayer' => 'requester',
                'SSECustomerAlgorithm' => '<string>',
                'SSECustomerKey' => '<string>',
                'SSECustomerKeyMD5' => '<string>',
                'SSEKMSEncryptionContext' => '<string>',
                'SSEKMSKeyId' => '<string>',
                'ServerSideEncryption' => 'AES256|aws:kms',
                'StorageClass' => 'STANDARD|REDUCED_REDUNDANCY|STANDARD_IA|ONEZONE_IA|INTELLIGENT_TIERING|GLACIER|DEEP_ARCHIVE|OUTPOSTS',
                'Tagging' => '<string>',
                'WebsiteRedirectLocation' => '<string>',
            ]);

        Called by Uppy on:

            https://github.com/transloadit/uppy/blob/master/packages/%40uppy/aws-s3-multipart/src/index.js#L78

            return this.client.post('s3/multipart', {
              filename: file.name,
              type: file.type,
              metadata
            }).then(assertServerError)

    */
    public function createMultipartUpload(Request $request)
    {
        $type = $request->input('type');
        $filenameRequest = $request->input('filename');
        $fileName = pathinfo($filenameRequest, PATHINFO_FILENAME);
        $fileExtension = pathinfo($filenameRequest, PATHINFO_EXTENSION);
        $folder = config('uppy-s3-multipart-upload.s3.bucket.folder') ? config('uppy-s3-multipart-upload.s3.bucket.folder').'/' : '';
        $key = $folder.Str::of($fileName.'_'.microtime())->slug('_').'.'.$fileExtension;

        try {
            $result = $this->client->createMultipartUpload([
                'Bucket'             => $this->bucket,
                'Key'                => $key,
                'ContentType'        => $type,
                'ContentDisposition' => 'inline',
            ]);
        } catch (Throwable $exception) {
            return response()
                ->json([
                    'message' => $exception->getMessage(),
                ], $exception->getStatusCode());
        }

        return response()
            ->json([
                'uploadId' => $result['UploadId'],
                'key'      => $result['Key'],
            ]);
    }

    /*
        List multipart upload parts.

        S3 Syntax:

            https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#listparts

            $result = $client->listParts([
                'Bucket' => '<string>', // REQUIRED
                'ExpectedBucketOwner' => '<string>',
                'Key' => '<string>', // REQUIRED
                'MaxParts' => <integer>,
                'PartNumberMarker' => <integer>,
                'RequestPayer' => 'requester',
                'UploadId' => '<string>', // REQUIRED
            ]);

        Called by Uppy on:

            https://github.com/transloadit/uppy/blob/master/packages/%40uppy/aws-s3-multipart/src/index.js#L96

            return this.client.get(`s3/multipart/${uploadId}?key=${filename}`)
            .then(assertServerError)
    */
    public function getUploadedParts(Request $request, $uploadId)
    {
        $key = $request->input('key');

        $parts = $this->listPartsPage($key, $uploadId, 0);

        return response()
            ->json($parts);
    }

    private function listPartsPage($key, $uploadId, $partIndex, $parts = null)
    {
        $parts = $parts ?? collect();

        $results = $this->client->listParts([
            'Bucket'           => $this->bucket,
            'Key'              => $key,
            'UploadId'         => $uploadId,
            'PartNumberMarker' => $partIndex,
        ]);

        if ($results['Parts']) {
            $parts = $parts->concat($results['Parts']);

            if ($results['IsTruncated']) {
                $results = $this->listPartsPage($key, $uploadId, $results['NextPartNumberMarker'], $parts);
                $parts = $parts->concat($results['Parts']);
            }
        }

        return $parts;
    }

    /*
        Generates a signed URL to upload a single part.

        S3 upload part:

            $result = $client->uploadPart([
                'Body' => <string || resource || Psr\Http\Message\StreamInterface>,
                'Bucket' => '<string>', // REQUIRED
                'ContentLength' => <integer>,
                'ContentSHA256' => '<string>',
                'ExpectedBucketOwner' => '<string>',
                'Key' => '<string>', // REQUIRED
                'PartNumber' => <integer>, // REQUIRED
                'RequestPayer' => 'requester',
                'SSECustomerAlgorithm' => '<string>',
                'SSECustomerKey' => '<string>',
                'SSECustomerKeyMD5' => '<string>',
                'SourceFile' => '<string>',
                'UploadId' => '<string>', // REQUIRED
            ]);

        Called by Uppy on:

            https://github.com/transloadit/uppy/blob/master/packages/%40uppy/aws-s3-multipart/src/index.js#L104

            return this.client.get(`s3/multipart/${uploadId}/batch?key=${filename}&partNumbers=${partNumbers.join(',')}`)
            .then(assertServerError)
    */
    public function prepareUploadParts(Request $request, $uploadId)
    {
        $key = $this->encodeURIComponent($request->input('key'));

        $partNumbers = explode(',', $request->input('partNumbers'));

        $presignedUrls = [];

        foreach ($partNumbers as $partNumber) {
            $command = $this->client->getCommand('uploadPart', [
                'Bucket'     => $this->bucket,
                'Key'        => $key,
                'UploadId'   => $uploadId,
                'PartNumber' => (int) $partNumber,
                'Body'       => '',
            ]);

            $presignedUrls[$partNumber] = (string) $this->client->createPresignedRequest($command, config('uppy-s3-multipart-upload.s3.presigned_url.expiry_time'))->getUri();
        }

        return response()
            ->json([
                'presignedUrls' => $presignedUrls,
            ]);
    }

    /*
        Completes a multipart upload by assembling previously uploaded parts.

        S3 Syntax:

            https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#completemultipartupload

            $result = $client->completeMultipartUpload([
                'Bucket' => '<string>', // REQUIRED
                'ExpectedBucketOwner' => '<string>',
                'Key' => '<string>', // REQUIRED
                'MultipartUpload' => [
                    'Parts' => [
                        [
                            'ETag' => '<string>',
                            'PartNumber' => <integer>,
                        ],
                        // ...
                    ],
                ],
                'RequestPayer' => 'requester',
                'UploadId' => '<string>', // REQUIRED
            ]);

        Called by Uppy on:

            https://github.com/transloadit/uppy/blob/master/packages/%40uppy/aws-s3-multipart/src/index.js#L112

            return this.client.post(`s3/multipart/${uploadIdEnc}/complete?key=${filename}`, { parts })
            .then(assertServerError)
    */
    public function completeMultipartUpload(Request $request, $uploadId)
    {
        $key = $this->encodeURIComponent($request->input('key'));

        $parts = $request->input('parts');

        // TODO: isValid Part
        // if (!Array.isArray(parts) || !parts.every(isValidPart)) {
        // return res.status(400).json({ error: 's3: `parts` must be an array of {ETag, PartNumber} objects.' })
        // }
        //
        // function isValidPart (part) {
        //     return part && typeof part === 'object' && typeof part.PartNumber === 'number' && typeof part.ETag === 'string'
        //}

        $result = $this->client->completeMultipartUpload([
            'Bucket'          => $this->bucket,
            'Key'             => $key,
            'UploadId'        => $this->encodeURIComponent($uploadId),
            'MultipartUpload' => ['Parts' => $parts],
        ]);

        $location = $result['Location'];

        return response()
            ->json([
                'location' => $location,
            ]);
    }

    /*
        Aborts a multipart upload.

        S3 Syntax:

            https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#abortmultipartupload
            $result = $client->abortMultipartUpload([
                'Bucket' => '<string>', // REQUIRED
                'ExpectedBucketOwner' => '<string>',
                'Key' => '<string>', // REQUIRED
                'RequestPayer' => 'requester',
                'UploadId' => '<string>', // REQUIRED
            ]);

        Called by Uppy on:

            https://github.com/transloadit/uppy/blob/master/packages/%40uppy/aws-s3-multipart/src/index.js#L121
            return this.client.delete(`s3/multipart/${uploadIdEnc}?key=${filename}`)
            .then(assertServerError)
    */
    public function abortMultipartUpload(Request $request, $uploadId)
    {
        $key = $request->input('key');

        $result = $this->client->abortMultipartUpload([
            'Bucket'   => $this->bucket,
            'Key'      => $this->encodeURIComponent($key),
            'UploadId' => $this->encodeURIComponent($uploadId),
        ]);

        return response()
            ->json([]);
    }
}
